File: sdk/java/src/main/java/com/aligntech/sdk/Evaluator.java
package com.aligntech.sdk;

import java.util.Collection;
import java.util.List;
import java.util.Map;


class Evaluator {

    EvaluationResult evaluate(FlagConfig flag, AlignTechUser user) {
        if (!flag.isActive()) {
            return new EvaluationResult(flag.getFlagKey(), false, null, null, "flag_inactive");
        }

        Map<String, Object> rollout = flag.getRollout();
        if (rollout.isEmpty()) {
            return new EvaluationResult(flag.getFlagKey(), false, null, null, "no_rollout_config");
        }

        String type = (String) rollout.getOrDefault("type", "boolean");

        boolean enabled = switch (type) {
            case "whitelist" -> {
                List<String> userIds = toStrList(rollout.get("userIds"));
                yield userIds != null && !userIds.isEmpty()
                        && user.getUserId() != null
                        && matchesAny(user.getUserId(), userIds);
            }
            case "boolean" -> Boolean.TRUE.equals(rollout.get("value"));
            default -> false;
        };

        return new EvaluationResult(flag.getFlagKey(), enabled, null, null, "rollout:" + type);
    }

    private boolean matchesAny(String value, List<String> patterns) {
        for (String pattern : patterns) {
            if (pattern.equals("*")) return true;
            if (pattern.startsWith("*") && pattern.endsWith("*"))
                return value.contains(pattern.substring(1, pattern.length() - 1));
            else if (pattern.startsWith("*"))
                return value.endsWith(pattern.substring(1));
            else if (pattern.endsWith("*"))
                return value.startsWith(pattern.substring(0, pattern.length() - 1));
            else if (pattern.equals(value)) return true;
        }
        return false;
    }

    @SuppressWarnings("unchecked")
    private static List<String> toStrList(Object obj) {
        if (obj instanceof List<?> list) return (List<String>) list;
        if (obj instanceof Collection<?> col) return List.copyOf((Collection<String>) col);
        return null;
    }

    record EvaluationResult(String flagKey, boolean enabled, String variant, Object payload, String reason) {}
}

File: sdk/swift/Sources/AlignTechFeatureSDK/Evaluator.swift
import Foundation

struct Evaluator {
    func evaluate(flag: FlagConfig, user: AlignTechUser) -> EvaluationResult {
        let base = { (enabled: Bool, reason: String) in
            EvaluationResult(
                flagKey: flag.flagKey,
                enabled: enabled,
                reason: reason
            )
        }
        
        guard flag.isActive else {
            return base(false, "flag_inactive")
        }
        
        guard let rollout = flag.rollout, !rollout.isEmpty else {
            return base(false, "no_rollout_config")
        }
        
        let type = (rollout["type"]?.value as? String) ?? "boolean"
        var enabled = false
        
        switch type {
        case "whitelist":
            if let userIds = rollout["userIds"]?.value as? [String],
               !userIds.isEmpty,
               let userId = user.userId {
                enabled = matchesAny(value: userId, patterns: userIds)
            }
            
        case "boolean":
            enabled = (rollout["value"]?.value as? Bool) ?? false
            
        default:
            enabled = false
        }
        
        return base(enabled, "rollout:\(type)")
    }
    
    private func matchesAny(value: String, patterns: [String]) -> Bool {
        for pattern in patterns {
            if pattern == "*" {
                return true
            }
            
            if pattern.hasPrefix("*") && pattern.hasSuffix("*") {
                let substr = String(pattern.dropFirst().dropLast())
                if value.contains(substr) {
                    return true
                }
            } else if pattern.hasPrefix("*") {
                let suffix = String(pattern.dropFirst())
                if value.hasSuffix(suffix) {
                    return true
                }
            } else if pattern.hasSuffix("*") {
                let prefix = String(pattern.dropLast())
                if value.hasPrefix(prefix) {
                    return true
                }
            } else if pattern == value {
                return true
            }
        }
        return false
    }
}
